File: scripts/find_project_sessions.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Find Project Sessions (Universal Scanner & Introspection Entrypoint)
Aggregates sessions across Antigravity, Codex, and Claude Code.
Supports --inspect to extract recent user prompts and touched files for AI tagging.
"""

import sys
import os
import json
import argparse
from pathlib import Path
import task_loop_state as state_store

# Add providers directory to sys.path
script_dir = Path(__file__).resolve().parent
providers_dir = script_dir / "providers"
sys.path.insert(0, str(providers_dir))

from get_agy_project_sessions import scan_agy_sessions
from get_codex_project_sessions import get_current_session_metadata, scan_codex_sessions
from get_claude_project_sessions import scan_claude_sessions
from get_zcode_project_sessions import scan_zcode_sessions


def safe_scan(vendor_name: str, scan_fn) -> list:
    try:
        return [
            {**item, "vendor": state_store.normalize_vendor(item.get("vendor")) or vendor_name}
            for item in (scan_fn() or [])
        ]
    except Exception as e:
        sys.stderr.write(f"[find-sessions] {vendor_name} provider error: {e}\n")
        return []


def find_sessions(project_root: str = ".", vendor: str = "Auto", inspect: bool = False) -> list:
    all_sessions = []
    vendor = state_store.normalize_vendor(vendor) or vendor.lower()

    scan_agy = lambda: safe_scan("antigravity", lambda: scan_agy_sessions(project_root, inspect_activity=inspect))
    scan_codex = lambda: safe_scan("codex", lambda: scan_codex_sessions(project_root))
    scan_claude = lambda: safe_scan("claude", lambda: scan_claude_sessions(project_root, inspect_activity=inspect))
    scan_zcode = lambda: safe_scan("zcode", lambda: scan_zcode_sessions(project_root, inspect_activity=inspect))

    if vendor == "antigravity":
        all_sessions.extend(scan_agy())
    elif vendor == "codex":
        all_sessions.extend(scan_codex())
    elif vendor == "claude":
        all_sessions.extend(scan_claude())
    elif vendor == "zcode":
        all_sessions.extend(scan_zcode())
    elif vendor == "all":
        all_sessions.extend(scan_agy())
        all_sessions.extend(scan_codex())
        all_sessions.extend(scan_claude())
        all_sessions.extend(scan_zcode())
    else:  # "auto"
        root_path = Path(project_root)
        policy_file = root_path / ".agents" / "task-loop" / "policy.json"
        if not policy_file.is_file():
            policy_file = root_path / ".codex" / "task-loop" / "policy.json"

        active_vendor = state_store.normalize_vendor(state_store.detect_vendor(os.environ)) or "antigravity"
        if policy_file.is_file():
            try:
                with open(policy_file, "r", encoding="utf-8") as f:
                    p = json.load(f)
                    if p.get("active_vendor"):
                        active_vendor = state_store.normalize_vendor(p["active_vendor"]) or active_vendor
            except Exception:
                pass
        primary_scans = {
            "antigravity": scan_agy,
            "codex": scan_codex,
            "claude": scan_claude,
            "zcode": scan_zcode,
        }
        primary = primary_scans.get(active_vendor, scan_agy)
        primary_sessions = primary()
        all_sessions.extend(primary_sessions)
        if not primary_sessions:
            for other_vendor, scan in primary_scans.items():
                if other_vendor != active_vendor:
                    all_sessions.extend(scan())

    unique = {}
    for session in all_sessions:
        identity = state_store.session_identity(session.get("vendor"), session.get("session_id"))
        if identity and identity not in unique:
            unique[identity] = session
    return sorted(unique.values(), key=lambda item: item.get("last_active_at") or "", reverse=True)


def main():
    if hasattr(sys.stdout, "reconfigure"):
        try:
            sys.stdout.reconfigure(encoding="utf-8")
        except Exception:
            pass
    parser = argparse.ArgumentParser(description="Find & Inspect project sessions across AI agents.")
    parser.add_argument("--root", default=".", help="Project root path (default: current directory)")
    parser.add_argument("--vendor", default="Auto", choices=["Auto", "Antigravity", "Codex", "Claude", "Zcode", "All"], help="Vendor selector")
    parser.add_argument("--inspect", "-i", action="store_true", help="Extract recent prompts and touched files for AI tagging")
    parser.add_argument("--current", action="store_true", help="Read the current Codex thread ID from host runtime metadata")
    args = parser.parse_args()

    if args.current:
        if args.vendor.lower() != "codex":
            parser.error("--current is currently supported only with --vendor Codex")
        print(json.dumps(get_current_session_metadata(), ensure_ascii=False, indent=2))
        return

    sessions = find_sessions(args.root, args.vendor, args.inspect)
    print(json.dumps(sessions, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/find_project_sessions.js
#!/usr/bin/env node
/**
 * Find Project Sessions (Universal Scanner & Introspection Entrypoint - Node.js)
 * Aggregates sessions across Antigravity, Codex, and Claude Code.
 */

const fs = require('fs');
const path = require('path');
const stateStore = require('./task_loop_state');

const { scanAgySessions } = require('./providers/get_agy_project_sessions');
const { scanCodexSessions, getCurrentSessionMetadata } = require('./providers/get_codex_project_sessions');
const { scanClaudeSessions } = require('./providers/get_claude_project_sessions');
const { scanZcodeSessions } = require('./providers/get_zcode_project_sessions');

function safeScan(vendorName, scanFn) {
  try {
    return (scanFn() || []).map(item => ({
      ...item,
      vendor: stateStore.normalizeVendor(item && item.vendor) || vendorName
    }));
  } catch (err) {
    console.error(`[find-sessions] ${vendorName} provider error: ${err.message}`);
    return [];
  }
}

function findSessions(projectRoot = '.', vendor = 'Auto', inspect = false) {
  let allSessions = [];
  const v = stateStore.normalizeVendor(vendor) || (vendor || 'auto').toLowerCase();

  const scanAgy = () => safeScan('antigravity', () => scanAgySessions(projectRoot, null, inspect));
  const scanCodex = () => safeScan('codex', () => scanCodexSessions(projectRoot));
  const scanClaude = () => safeScan('claude', () => scanClaudeSessions(projectRoot, null, inspect));
  const scanZcode = () => safeScan('zcode', () => scanZcodeSessions(projectRoot, null, inspect));

  if (v === 'antigravity') {
    allSessions = allSessions.concat(scanAgy());
  } else if (v === 'codex') {
    allSessions = allSessions.concat(scanCodex());
  } else if (v === 'claude') {
    allSessions = allSessions.concat(scanClaude());
  } else if (v === 'zcode') {
    allSessions = allSessions.concat(scanZcode());
  } else if (v === 'all') {
    allSessions = allSessions.concat(scanAgy());
    allSessions = allSessions.concat(scanCodex());
    allSessions = allSessions.concat(scanClaude());
    allSessions = allSessions.concat(scanZcode());
  } else { // 'auto'
    const rootPath = path.resolve(projectRoot);
    let policyFile = path.join(rootPath, '.agents', 'task-loop', 'policy.json');
    if (!fs.existsSync(policyFile)) {
      policyFile = path.join(rootPath, '.codex', 'task-loop', 'policy.json');
    }

    let activeVendor = stateStore.normalizeVendor(stateStore.detectVendor(process.env)) || 'antigravity';
    if (fs.existsSync(policyFile)) {
      try {
        const p = JSON.parse(fs.readFileSync(policyFile, 'utf8'));
        if (p.active_vendor) {
          activeVendor = stateStore.normalizeVendor(p.active_vendor) || activeVendor;
        }
      } catch (e) {}
    }
    const primaryScans = { antigravity: scanAgy, codex: scanCodex, claude: scanClaude, zcode: scanZcode };
    const primary = primaryScans[activeVendor] || scanAgy;
    const primarySessions = primary();
    allSessions = allSessions.concat(primarySessions);
    if (primarySessions.length === 0) {
      for (const [otherVendor, scan] of Object.entries(primaryScans)) {
        if (otherVendor !== activeVendor) allSessions = allSessions.concat(scan());
      }
    }
  }

  const unique = new Map();
  for (const session of allSessions) {
    const identity = stateStore.sessionIdentity(session.vendor, session.session_id);
    if (identity && !unique.has(identity)) unique.set(identity, session);
  }
  return Array.from(unique.values()).sort((a, b) => (b.last_active_at || '').localeCompare(a.last_active_at || ''));
}

function main() {
  const args = process.argv.slice(2);
  let projectRoot = '.';
  let vendor = 'Auto';
  let inspect = false;
  let current = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--root' && args[i + 1]) {
      projectRoot = args[++i];
    } else if (args[i] === '--vendor' && args[i + 1]) {
      vendor = args[++i];
    } else if (args[i] === '--inspect' || args[i] === '-i') {
      inspect = true;
    } else if (args[i] === '--current') {
      current = true;
    } else if (!args[i].startsWith('-')) {
      projectRoot = args[i];
    }
  }

  if (current) {
    if (vendor.toLowerCase() !== 'codex') {
      console.error('--current is currently supported only with --vendor Codex');
      process.exit(1);
    }
    console.log(JSON.stringify(getCurrentSessionMetadata(), null, 2));
    return;
  }

  const sessions = findSessions(projectRoot, vendor, inspect);
  console.log(JSON.stringify(sessions, null, 2));
}

if (require.main === module) {
  main();
}

module.exports = { findSessions };
